crates/app/src/app.rs

```diff
@@ -0,0 +1,190 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{repos, tags, trees};
+
+use std::fmt::Display;
+use std::sync::Arc;
+
+use drawbridge_store::{Get, Memory};
+use drawbridge_type::{repository, tag, tree};
+
+use axum::body::Body;
+use axum::handler::Handler;
+use axum::http::{Method, Request, StatusCode};
+use axum::response::IntoResponse;
+use axum::routing::IntoMakeService;
+use axum::{Extension, Router};
+use tokio::sync::RwLock;
+use tower::Service;
+
+pub(crate) type RepoStore = RwLock<Memory<repository::Name>>;
+pub(crate) type TagStore = RwLock<Memory<(repository::Name, tag::Name)>>;
+pub(crate) type TreeStore = RwLock<Memory<(repository::Name, tag::Name, tree::Path)>>;
+
+pub(crate) async fn assert_repo(
+    repos: Arc<RepoStore>,
+    repo: repository::Name,
+) -> Result<(), (StatusCode, &'static str)> {
+    #[inline]
+    fn err_map(e: impl Display) -> (StatusCode, &'static str) {
+        eprintln!("failed to check repository existence: {}", e);
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Failed to check repository existence",
+        )
+    }
+    if !repos
+        .read()
+        .await
+        .contains(repo.clone())
+        .await
+        .map_err(err_map)?
+    {
+        Err((StatusCode::NOT_FOUND, "Repository does not exist"))
+    } else {
+        Ok(())
+    }
+}
+
+pub(crate) async fn assert_tag(
+    tags: Arc<TagStore>,
+    repo: repository::Name,
+    tag: tag::Name,
+) -> Result<(), (StatusCode, &'static str)> {
+    #[inline]
+    fn err_map(e: impl Display) -> (StatusCode, &'static str) {
+        eprintln!("failed to check tag existence: {}", e);
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Failed to check tag existence",
+        )
+    }
+    if !tags
+        .read()
+        .await
+        .contains((repo.clone(), tag.clone()))
+        .await
+        .map_err(err_map)?
+    {
+        Err((StatusCode::NOT_FOUND, "Tag does not exist"))
+    } else {
+        Ok(())
+    }
+}
+
+/// Parses the URI of `req` and routes it to respective component.
+async fn handle(mut req: Request<Body>) -> impl IntoResponse {
+    let path = req.uri().path().strip_prefix('/').expect("invalid URI");
+    let (head, tail) = path
+        .split_once("/_")
+        .map(|(left, right)| (left, format!("_{}", right)))
+        .unwrap_or((path, "".into()));
+    if head.is_empty() {
+        return Err((
+            StatusCode::NOT_FOUND,
+            format!("Route `/{}` not found", path),
+        ));
+    }
+    let repo = head.parse::<repository::Name>().map_err(|e| {
+        (
+            StatusCode::BAD_REQUEST,
+            format!("Failed to parse repository name: {}", e),
+        )
+    })?;
+    assert_eq!(
```

Comment:
Is it kosher to have assertions in non-test functions?

Reply:
Actually, yes, they are used widely in stdlib.
There's also `debug_assert_*` for performance-sensitive applications, but in this case cost of doing this is minuscule compared to network I/O.
Since it's a implementation error, and a big one, we *want* to panic.